File: Desktop/MotivateMe/src/util/DataAccess.java
package util;

import javafx.collections.FXCollections;
import javafx.collections.ObservableList;

import model.Quote;

import org.json.JSONArray;
import org.json.JSONObject;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;


public class DataAccess {

    private static HttpClient httpClient ;


    public static Quote getRandomQuote() {

        Quote quote ;

        httpClient = HttpClient.newBuilder()
                .version(HttpClient.Version.HTTP_2)
                .connectTimeout(Duration.ofSeconds(10))
                .build() ;

        try {
            HttpRequest request = HttpRequest.newBuilder()
                    .GET()
                    .uri(URI.create("http://localhost:8080/motivate/me/quote"))
                    .build() ;

            HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString()) ;

            // convert the body of the HttpResponse to JSON
            JSONObject jsonObject = new JSONObject(response.body()) ;


            // extract the attributes of the Quote from the JSON
            quote = new Quote(jsonObject.getString("id"), jsonObject.getString("quotation"), jsonObject.getString("author"),  jsonObject.getString("category")) ;
        }
        catch (Exception e) {
            System.out.println("\nExceptions in DataAccess.getRandomQuote():\n" + e.getMessage() + "\n") ;
            quote = new Quote("ERROR", e.getMessage(), "ERROR", "ERROR") ;
        }

        return quote ;
    }


    public static ObservableList<Quote> getAllQuotes() {

        ObservableList<Quote> quotesList = FXCollections.observableArrayList() ;

        httpClient = HttpClient.newBuilder()
                .version(HttpClient.Version.HTTP_2)
                .connectTimeout(Duration.ofSeconds(10))
                .build() ;

        try {
            HttpRequest request = HttpRequest.newBuilder()
                    .GET()
                    .uri(URI.create("http://localhost:8080/motivate/me/quotes/all"))
                    .build() ;

            HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString()) ;


            JSONArray array = new JSONArray(response.body()) ;

            for(int i = 0 ; i < array.length() ; i++) {
                JSONObject jsonObject = array.getJSONObject(i) ;

                quotesList.add(new Quote(jsonObject.getString("id"), jsonObject.getString("quotation"), jsonObject.getString("author"),  jsonObject.getString("category"))) ;
            }
        }
        catch (Exception e) {
            System.out.println("\nExceptions in  DataAccess.getAllQuotes():\n" + e.getMessage() + "\n") ;

            quotesList.clear() ;
            quotesList.add(new Quote("ERROR", e.getMessage(), "ERROR", "ERROR")) ;
        }

        return quotesList ;
    }


    /**
     * Used to obtain all the quotes by a specific author
     * @param nameOfAuthor the name of the desired author
     * @return a list of Quotes
     */
    public static ObservableList<Quote> getQuotesFrom(String nameOfAuthor) {

        ObservableList<Quote> quotesList = FXCollections.observableArrayList() ;

        httpClient = HttpClient.newBuilder()
                .version(HttpClient.Version.HTTP_2)
                .connectTimeout(Duration.ofSeconds(10))
                .build() ;

        nameOfAuthor = nameOfAuthor.replaceAll(" ", "%20") ;

        try {
            HttpRequest request = HttpRequest.newBuilder()
                    .GET()
                    .uri(URI.create("http://localhost:8080/motivate/me/quote/author/all/?author=" + nameOfAuthor))
                    .build() ;

            HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString()) ;


            JSONArray array = new JSONArray(response.body()) ;

            for(int i = 0 ; i < array.length() ; i++) {
                JSONObject jsonObject = array.getJSONObject(i) ;

                quotesList.add(new Quote(jsonObject.getString("id"), jsonObject.getString("quotation"), jsonObject.getString("author"),  jsonObject.getString("category"))) ;
            }
        }
        catch (Exception e) {
            System.out.println("\nExceptions in  DataAccess.getQuotesFrom(String nameOfAuthor):\n" + e.getMessage() + "\n") ;

            quotesList.clear() ;
            quotesList.add(new Quote("ERROR", e.getMessage(), "ERROR", "ERROR")) ;
        }

        return quotesList ;
    }


    public static ObservableList<Quote> getQuotesOn(String category) {

        ObservableList<Quote> quotesList = FXCollections.observableArrayList() ;

        httpClient = HttpClient.newBuilder()
                .version(HttpClient.Version.HTTP_2)
                .connectTimeout(Duration.ofSeconds(10))
                .build() ;

        try {
            HttpRequest request = HttpRequest.newBuilder()
                    .GET()
                    .uri(URI.create("http://localhost:8080/motivate/me/quote/category/all/?category=" + category))
                    .build() ;

            HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString()) ;


            JSONArray array = new JSONArray(response.body()) ;

            for(int i = 0 ; i < array.length() ; i++) {
                JSONObject jsonObject = array.getJSONObject(i) ;

                quotesList.add(new Quote(jsonObject.getString("id"), jsonObject.getString("quotation"), jsonObject.getString("author"),  jsonObject.getString("category"))) ;
            }
        }
        catch (Exception e) {
            System.out.println("\nExceptions in  DataAccess.getQuotesOn(String category):\n" + e.getMessage() + "\n") ;

            quotesList.clear() ;
            quotesList.add(new Quote("ERROR", e.getMessage(), "ERROR", "ERROR")) ;
        }

        return quotesList ;
    }


    public static Quote getQuoteWith(String id) {
        Quote quote ;

        httpClient = HttpClient.newBuilder()
                .version(HttpClient.Version.HTTP_2)
                .connectTimeout(Duration.ofSeconds(10))
                .build() ;

        try {
            HttpRequest request = HttpRequest.newBuilder()
                    .GET()
                    .uri(URI.create("http://localhost:8080/motivate/me/quote/id/?id=" + id))
                    .build() ;

            HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString()) ;

            JSONObject jsonObject = new JSONObject(response.body()) ;

            quote = new Quote(jsonObject.getString("id"), jsonObject.getString("quotation"), jsonObject.getString("author"),  jsonObject.getString("category")) ;
        }
        catch (Exception e) {
            quote = new Quote("ERROR", e.getMessage(), "ERROR", "ERROR") ;
            System.out.println("\nExceptions in  DataAccess.getQuoteWith(String id):\n" + e.getMessage() + "\n") ;
        }

        return quote ;
    }
}

File: Mobile/motivate_me/lib/util/data_access.dart
import 'dart:convert';
import 'dart:math';

import 'package:motivate_me/util/quote.dart';
import 'package:http/http.dart' ;

class DataAccess {
  static List<Quote> quotesList = [] ;

  static Future<List<Quote>> getAllQuotes() async {
    
    if (quotesList.isEmpty) {
      quotesList = await _getAllQuotesFromServer() ;
    }

    return quotesList ;
  }

  static Future<List<Quote>> _getAllQuotesFromServer() async {
  
    try {
      Response response = await get(Uri.parse("http://10.0.2.2:8080/motivate/me/quotes/all")).timeout(const Duration(seconds: 15)) ;

      List dataList = jsonDecode(response.body) ;

      for (int i = 0 ; i < dataList.length ; i++) {
        Map data = dataList[i] ;
        Quote quote = Quote(id: data['id'], quotation: data['quotation'], author: data['author'], category: data['category']) ;
        quotesList.add(quote) ;
      }

      print("Quotes obtained successfully!") ;
    }
    catch (e) {
      print("\nExceptions in DataAccess.getAllQuotes() -> ") ;
      print(e.toString()) ;

      quotesList.clear() ;
      quotesList.add(Quote(id: "ERROR", quotation: e.toString(), author: "ERROR", category: "ERROR")) ;
    }

    return quotesList ;
  }


  static Quote getRandomQuote() {
    int n = Random().nextInt(quotesList.length) ;
    return quotesList[n] ;
  }


  static List<Quote> getQuotesBy(String author) {
    List<Quote> results = quotesList.where((element) => element.author.toLowerCase().contains(author.toLowerCase())).toList() ;

    if (results.isEmpty) {
      results.add(Quote(id: "-", quotation: "<< no qoute by $author found! >>", author: "-", category: "-")) ;
    }

    return results ;
  }


  static List<Quote> getQuotesOn(String category) {
    List<Quote> results = quotesList.where((element) => element.category.toLowerCase().contains(category.toLowerCase())).toList() ;

    if(results.isEmpty) {
      results.add(Quote(id: "-", quotation: "<< no quote on $category found! >>", author: "-", category: "-")) ;
    }

    return results ;
  }


  static Quote getQuote(String id) {

    var results = quotesList.where((element) => element.id == id) ;

    if (results.isEmpty) {
      return Quote(id: "", quotation: "<< no quote found with id $id", author: "", category: "") ;
    }

    return results.first ;
  }
}